Refactor: make controller modular

NuttyShrimp opened the pull request:
- Rewritten the controller to a class based package
- Moved code depending on third-party libraries to their separate class, so these are swappable with custom one's

## How do I test this PR?
- Select a file (I've mostly used the README file)
- In the right column, click the "Edit" button. This should open a drawer. 
  This drawer should contain a table with all the subjects who have access to the file
- Click the "New Subject" to give a new subject access. This will again open a new smaller drawer where you can select the subject type
- Select "WebId" and give a webId which already has been added to the ACL file/table (e.g. your own webId)
- This should give an error in the console + a toast notification that the webId has been added previously
- Now use a webId which isn't present in the ACL file/table
- Click on the "Create" button, the Drawer should close & the new webId should be visible in the table
- Now choose one of the entries in the table & hit the "Edit" button at the end of the row
- Grant a new permission by checking the checkbox in the new drawer that should've opened.
- When you close this drawer. You'll see that your changes are reflected in the table (give it a second to reload the data)
- You can do multiple grants/revokes at once in the "edit subject" drawer
  - If you do them too fast after each other the inrupt SDK will freak out and skip some changes to the ACL file which will result in a desync between the ACL & the index.json


depends on #16 
closes #8 
Does a bit of work for #19, the resource access is removed when all the grants are revoked

NuttyShrimp committed:
feat: add select with creation option for granting new users permissions

NuttyShrimp committed:
refactor(controller): split logic in more file to better understand controller + apply suggested index entry format

NuttyShrimp committed:
refactor(controller): layout structure for state-based access-management

NuttyShrimp committed:
feat: expand generic usage for stricter types around the Subject

NuttyShrimp committed:
feat(types): finish abstracting modules

NuttyShrimp committed:
feat(accessManagement): implement add+create permission

NuttyShrimp committed:
feat(stores): always get a reference to index/item

NuttyShrimp committed:
refactor(builder): fix Subject(Resolver) types for building

NuttyShrimp committed:
feat(accessManagement): implement enable/disable permissions

NuttyShrimp committed:
feat(permissionManager): add missing function to get permissions from resources in a container

NuttyShrimp committed:
chore: remove old non-modulair file

NuttyShrimp committed:
feat(accessManagement): fix types to make implict definition of T working

NuttyShrimp committed:
fix(permissionManager): fixing broken permissionManager types

NuttyShrimp committed:
chore: cleanup builder restants + fix inruptController file

NuttyShrimp committed:
feat(controller): add label function for subjectResolvers

NuttyShrimp committed:
fix(controller): re-add updated index in root of controller project

NuttyShrimp committed:
feat(loama): cleanup + using new controller where possible

NuttyShrimp committed:
fix(controller): use filter to prevent duplicate entries when converting access modes to permissions

NuttyShrimp committed:
feat: replace subject select with fancier table & nested drawers

Seems like the changes made are not yet pushed to the index.json

NuttyShrimp committed:
chore(controller): rename AccessManagement to Controller

NuttyShrimp committed:
fix(loama): set right type to updatePerms parameter

NuttyShrimp committed:
feat(loama): disable append when write is granted

NuttyShrimp committed:
feat(loama+controller): add cache busting so table can be refreshed when permission editor is closed

NuttyShrimp committed:
chore: add missing types + add proper dev server for controller

NuttyShrimp committed:
fix(controller: use cached index when retrieving an item

NuttyShrimp committed:
fix(loama): add permissionChangeHandler to control switch

NuttyShrimp committed:
feat(controller): introduce mutex to prevent accidental overwrites

NuttyShrimp committed:
feat(permissionDrawer): disable checkboxes when update is pending

NuttyShrimp commented:
I've come across #22 on this branch, but do find this issue out-of-scope for this PR, as it was probably already an issue before the rewrite

bjdmeest commented:
Can you resolve conflicts? @NuttyShrimp

NuttyShrimp committed:
Merge branch 'development' into refactor/controller-strategy-pattern

NuttyShrimp committed:
docs(changelog): add entry for PR

bjdmeest commented on controller/src/classes/stores/InruptStore.ts:
```diff
@@ -0,0 +1,65 @@
+import { BaseStore } from "./BaseStore";
+import { getDefaultSession, Session } from "@inrupt/solid-client-authn-browser";
+
+export class InruptStore extends BaseStore implements IStore {
```
I don't see a reason for an InruptStore. It should be a store independent of the permissionmanager

NuttyShrimp replied:
The inrupt prefix is to indicate it uses the inrupt SDK to interact with your pod.
You could replace it with another store and still be able to use `InruptPermissionManager` & vice versa

bjdmeest replied:
Ah OK, that wasn't clear, maybe just add a small comment for this class to clarify.

NuttyShrimp replied:
Added it in 6008dd8

bjdmeest commented on controller/src/classes/Controller.ts:
```diff
@@ -0,0 +1,165 @@
+    constructor(store: IStore, subjectResolvers: EnforceKeyMatchResolver<T>, pm: IPermissionManager<T>) {
+        super();
+        this.store = store;
+        this.permissionManager = pm;
```
I would expect, for each 'permission type' to have a permission function --> you could have a webId permission function (eg from inrupt), and a qualifier permission function (e.g. from KNoWS)

bjdmeest commented on controller/src/classes/subjectResolvers/Public.ts:
```diff
@@ -0,0 +1,22 @@
+import { PublicSubject } from "../../types/subjects";
+import { ISubjectResolver } from "../../types/modules";
+
+export class PublicResolver implements ISubjectResolver<PublicSubject> {
```
very good to have separate resolvers, independent of the permission manager

bjdmeest commented on controller/src/inruptController.ts:
```diff
@@ -0,0 +1,26 @@
+// )
+
+export const inruptController = new Controller(
+    new InruptStore(),
```
yeah, so I would expect rather something like::

```
export const inruptController = new Controller(
    new Store(),
    {
        webId: {
            resolver: new WebIdResolver(),
            manager: new InruptPermissionManager.webIdManager()
        }, 
        public: {[same for other types]}
    })
```

NuttyShrimp replied:
So the `webIdManager`should include each function the current `IPermissionManager` interface defines? Or do you have anything else in mind?

bjdmeest replied:
Yes, because the WebIdResolver might use a different location to lookup/store the permissions (eg ACL files) than the qualifierResolver (eg ACP files).

bjdmeest replied:
right now it's assumed that one permissionmanager will handle all permission types, but in reality it'll be 2-3 permissionmanagers, each handling their own (disjoint) set of permission types

bjdmeest commented on controller/src/index.ts:
```diff
@@ -1,2 +1,5 @@
-export * from "./pod";
-export * from "./access-management"
+// TODO: move this file to the types folder
```
if you already have a todo in such a large MR, why no just do it here?

NuttyShrimp committed:
chore(controller): move types file into designated folder

bjdmeest commented:
The 'new subject' button doesn't work

I suggest to add a scenario description of what you should try out to showcase all functionalities. As long as we don't have automated testing, we can follow that scenario to 'test'

bjdmeest commented:
When adding write access to, e.g. the profile document, and then later revoking that write access, the revocation is not permanent

NuttyShrimp committed:
chore(controller): add comment to elaborate on the "Inrupt" prefix

NuttyShrimp commented:
> When adding write access to, e.g. the profile document, and then later revoking that write access, the revocation is not permanent

I think this is related to #22 and/or #1. But I don't have a direct solution to this problem yet. I'm going to investigate it after the other comments are resolved

NuttyShrimp committed:
feat(loama): add drawer behind new subject button

NuttyShrimp committed:
refactor(controller): add a manager per subject type

NuttyShrimp committed:
fix(permissionManager): remove use of "difference" function

This API is way to new to use w/o polyfill. But core-js was refusing to
work with tsc

NuttyShrimp committed:
fix(loama): add check if the webId field actually contains data

NuttyShrimp commented:
> The 'new subject' button doesn't work
> 
> I suggest to add a scenario description of what you should try out to showcase all functionalities. As long as we don't have automated testing, we can follow that scenario to 'test'

The new subject button has gotten functionality & I've added a guideline how you can test everything that has been changed in loama

NuttyShrimp committed:
feat(loama): add toast for better ux on errors

NuttyShrimp committed:
fix(loama): add missing implementation for public in subjectForm

NuttyShrimp committed:
fix(controller): add current type to subjectSelector in webeId resolver

NuttyShrimp committed:
feat(controller): cleanup index item if SDK auto removes entry on revoke off all permissions

NuttyShrimp committed:
feat(controller): expose indexItem generic through index & related functions

NuttyShrimp committed:
feat(controller): add a double-check mechanism after updating permissions

This will decrease the change of the index going out of sync with the
ACL file

NuttyShrimp committed:
feat(loama): force wait until a permission update operation is over before closing drawer

NuttyShrimp committed:
fix: add explicit types to inruptController + add missed return value

bjdmeest commented on controller/src/inruptController.ts:
```diff
@@ -0,0 +1,24 @@
+import { WebIdResolver } from "./classes/subjectResolvers/WebId";
+import { PublicSubject, WebIdSubject } from "./types/subjects";
+
+export const inruptController = new Controller<{
```
minor: i'd just call this the controller (or ActiveController, or ConfiguredController, or whatever). It happens to use only inrupt-dependent functions, but that's 'by accident': if new (non-inrupt) functions get added, we'll change this file

NuttyShrimp committed:
chore: rename configured controllers file & instance

bjdmeest commented on CHANGELOG.md:
you'll probably wanna add the toaster

bjdmeest commented:
Code is good! Now just some documentation

Could you add both in the loama app readme and in the loama controller readme some documentation of the code? ie how the structure of the different permission subjects/handlers are, and how this can be instantiated in the activeController?

Also put the 'guideline how you can test everything that has been changed in loama' to documentation

NuttyShrimp committed:
chore(controller): export classes

This will allow people to create their own controllers with modules that
we've created

NuttyShrimp committed:
docs: extend the docs for controller, add manual test instruction to loama & update the changelog
